### ISSUE TITLE
Epic: Snapshots

### ISSUE THREAD

**Issue by jandubois** [rancher-sandbox/rancher-desktop#4236]
A snapshot stores the current configuration of the VM and all associated settings.

## Use cases

* A project requires a various base services to be installed and configured. Whenever the developer finishes a story, they want to revert back to this base configuration before starting work on the next feature.

* A developer is testing data migration. After a bug is detected and fixed, they want to start over with the original data and run the migration again.

* Similarly upgrading Kubernetes versions with an existing workload may need to be repeated multiple times with just different configurations, to verify that everything continues running after the upgrade. So it is necessary to revert back to the starting state again.

## Features

Snapshots will be given a name and stored in a snapshot directory. Available functions will be

* list available snapshots
* save a snapshot
* restore a snapshot
* delete a snapshot
* rename a snapshot
* prune incomplete snapshots to reclaim the disk space

These functions will be implemented by an `rdctl snapshot` subcommand. The VM must be stopped before a snapshot can be saved or restored.

The functionality will be exposed through the UI as well.

## Snapshot directory

Snapshots are stored in a `snapshots" directory:

* macOS: `~/Library/Application Support/rancher-desktop/snapshots`
* Linux: `~/.local/share/rancher-desktop/snapshots`
* Windows: `%LOCALAPPDATA%\rancher-desktop\snapshots`

On macOS the snapshot directory should be excluded from the Time Machine backup list at the time it is created. The user can remove the exclusion if they really want to include it.

The directory may be moved to a different location by creating a symlink in the default location.

## How it works internally

All the features are implemented as rdctl snapshot subcommands.

When necessary rdctl will send an API request to shutdown (and lock) the VM, or to unlock it and start it back up. The UI will display a modal dialog while the VM is stopped/locked to prevent the user from changing settings or starting the VM again too early.

rdctl uses another API to poll the backend state, to determine if the VM is running or stopped.

The stopped/locked state is also recorded in the app data directory. This allows the app to recognize during startup when a snapshot create/restore initiated from the CLI is already in progress. In that case a dialog will explain the situation, and the app quits when the dialog is dismissed.

There needs to be an rdctl unlock command to manually remove the lock in case rdctl dies while the VM is in a locked state.

The Rancher Desktop UI can also spawn the corresponding rdctl snapshot ... --json commands through a new API in the backend. The backend will keep reading STDOUT from the subprocess to get progress updates and error messages. These will be posted as events back to the UI. rdctl snapshot ... will still send the VM shutdown commands to the backend, so the UI doesn't need to do anything beyond launching the commands.

Each snapshot has a "this snapshot has completed successfully" marker in the filesystem. Snapshots without the marker will not be returned by rdctl snapshot ls. This marker is the first thing that gets deleted by rdctl snapshot delete.

When Rancher Desktop starts up and is not locked, it should run rdctl snapshot prune to look for any abandoned snapshots and delete them (with feedback to the user).

Restoring a snapshot should effectively be a factory-reset, followed by copying the database and settings back into place. There should be a marker that this operation is in progress, or if it has failed. When it has failed (and Rancher Desktop is not already running) then at the next start RD needs to detect this and inform the user.

### Backend

* [x] #5115
* [x] #5116
* [x] #5117
* [x] #5476
* [x] #5119
* [x] #5569 
* [x] #5570 
* [x] #5571
* [x] #5638 
* [x] #5626 
* [x] #5674
* [x] #5684
* [x] #5690
* [x] #5691

### Frontend

* [x] #4527
* [x] #5670 
* [x] #5677
* [x] #5593
* [x] #5748
* [x] #5596 
* [x] #4598
* [x] #5804

### QA/Docs

* [x] #5597
* [ ] Write BATS tests for the backend

### Later

* [ ] #5643  
* [ ] #5683
* [ ] #5555
* [ ] #5556
* [ ] #4597
* [ ] #5557
* [ ] #5627
* [ ] #5675
* [x] #5580 
* [ ] #5581 
* [ ] #5594 
* [ ] #5558 
* [ ] #5675 
* [ ] #5676
* [ ] #5706
* [ ] #5728
* [ ] #5736
* [ ] #5747
* [ ] #5751
* [ ] #5766
* [ ] #4599

## Acceptance criteria

- Snapshots should not be deleted on factory reset.
- Restoring a snapshot should restore the following things:
  - settings
  - kubernetes resources
  - docker/nerdctl running/stopped containers, images
  - override.yaml (for linux and macOS)
  - current state of the VM, 
  - disk
  - WONT INCLUDE DEPLOYMENT PROFILES
- Snapshots must have Name and creation date
- To save and store a snapshot you MUST stop the VM before saving it
- Be implemented via `rdctl snapshot` command
- Snapshots must be saved in the sanpshots directory in default location created by us. User can create a symlink to access the snapshots in their desired directory
- Directory must be in the same volume where the VM is stored (lima filesystem images). It’s an internal disk.
- Just user data will be saved.
- If there's a Locked profile different from the snapshots, the profile will override the snapshots when the VM restarts.
- RD should support snapshots from previous versions. RD must migrate the data from snapshots to the newer version
- Restoring the snapshot: Throw away the disk image from the VM and restart, NOT reset the VM
- Don’t support sharing snapshots between users nor between platforms
- Local volumes inside the VM must be PRESERVED. Stopped containers must not be removed from the VM.
- Progress bar in the UI to show the download progress of the snapshot being saved
- UI must show an Error message if it fails to create snapshots and message of success when they’re successfully saved
- - When taking a snapshot via UI, user cannot take a snapshot at the same time via CLI and vice-versa
- UI components must be locked while a snapshot operation is in progress
- If a snapshot is created/restored when app is not running, when starting the application, UI must be locked and needs to inform that the snapshot operation is in progress and the VM is still to start.
- `Rdctl create/restore snapshot` WILL PAUSE and restart THE VM for the user. 
- UI must show a list of saved snapshots displaying its name and date with 2 action buttons to restore and delete
- Tray menu options must be disabled while snapshot is in progress
- Once a snapshot operation starts, the preferences dialog must be closed
- By deleting a snapshot, the UI must remove it from the list of snapshots and `rdctl snapshot list` must not show the deleted snapshot anymore
- If locked deployment profile is implemented after a snapshot is created with different values of locked fields, the restore operation must be successful, but the locked values in the locked profile must override the ones from the snapshot

**Comment by jandubois** [comment]
> Do snapshots have any other attributes than date/time?

As discussed in the meeting today, snapshots should have a name as well, so you refer to them from the cli (e.g. in scripts):

```
rdctl snapshot save after-migration
rdctl snapshot restore base-config
```

I could see having a more verbose description as well, but let's start with just a name (and date/time).

**Comment by jandubois** [comment]
I updated the epic with:

> Snapshots are stored in a "Rancher Desktop Snapshots" directory under `%USERPROFILE%` on Windows or `$HOME` on Linux/macOS.

This is the same location where VirtualBox stores it's `VirtualBox VMs` directory for virtual machines.

This (`$HOME/Virtual Machines`) is also used by VMware Fusion 11 and later on macOS. Earlier Fusion versions use `$HOME/Documents/Virtual Machines`. VMware workstation on Linux uses `$HOME/vmware`.

We should not use the `Documents` location, at least on macOS, because it may be mapped to iCloud, or placed on Dropbox, which would not work very well at all.

For Linux we have an alternative candidate at `$HOME/.local/share/rancher-desktop-snapshots`. I somewhat prefer to use `$HOME/Rancher Desktop Snapshots` because of the symmetry with the other platforms.

I understand that some user might not like the clutter, but I think hiding away a directory that potentially includes several 100GB of data might not be wise either.

As a compromise I could see that we would use `$HOME/.local/share/rancher-desktop-snapshots` if it already exists, but otherwise use the default location.

**Comment by jandubois** [comment]
I did some experiments (on a macOS iMac) to get a feeling for the time needed to take/restore a snapshot.

I filled the data volume up to about 60GB:

```
$ du -h diffdisk
 61G	diffdisk
```

Copying using `clonefile` is instantaneous, as expected:

```
$ time cp -c diffdisk diffdisk2
real	0m0.028s
```

Doing a local copy (on the built-in SSD) takes 3½ minutes:

```
$ time cp diffdisk diffdisk3
real	3m33.496s
```

Repeating this again later took 4m13s, maybe because the disk is getting quite full.

Copying to an SSD RAID array connected over Thunderbolt was surprisingly fast:

```
$ time cp diffdisk /Volumes/ThunderBlade/
real	0m55.730s
```

Wheras copying over USB3 to a spinning rust disk was expectedly slow (even though not quite as slow as I expected based on the local copy time):

```
$ time cp diffdisk /Volumes/BladeDupe/
real	10m46.301s
```

I'm still surprised how slow reading from and writing to the same SSD was; I expected about the same performance as writing to the external array. It is probably faster because it is a 4 disk RAID0 array, so writes can go over multiple PCI channels in parallel. I should probably redo the experiment on a different machine, but for now these numbers are good enough to give us a feel for expected performance.

It looks like we'll definitely need a status dialog to allow the user to cancel the operation (and ideally show the progress). Because on Windows we don't have a copy-on-write option, and on Linux it will only be available on btrfs and maybe XFS. And even on macOS we can't do `clonefile` if this is a cross-device copy.

**Comment by adamkpickering** [comment]
> Snapshots are stored in a "Rancher Desktop Snapshots" directory under %USERPROFILE% on Windows or $HOME on Linux/macOS.

Two points:
1. [We should avoid spaces in directory names](https://superuser.com/questions/29111/what-technical-reasons-exist-for-not-using-space-characters-in-file-names). Names that contain spaces may cause problems, whereas names that don't contain spaces will never cause problems (at least, not due to there being spaces).
2. On Linux, I would be very surprised by the appearance of a `~/Rancher Desktop Snapshots` directory. I've never seen an application do this, and I'd look for an alternative if there was no way to change this. I am one of those users that hates having a polluted home directory. Even `~/.rancher-desktop-snapshots` would be better. Maybe this is typical on macOS, and maybe we should do what you suggest there, but it would be very weird on Linux.

---

Can you help me understand why it is important that we don't hide a directory that may contain a lot of data?

When I want to free up space on my disk, I use a tool for figuring out what directories contain the most data, for example `du -d 1 .`. There are graphical tools that do the same thing. If using a tool like this it doesn't matter whether the directory is hidden - it gets found by the tool.

Also, the `diffdisk` that is in use is already in a hidden directory, and it could be very big.

I just don't see the benefit of looking at snapshots as user data when the user will never interact directly with the files. In the future we may want to provide users with the ability to share snapshots with other users, but in that case we would have to make them into tgz/zip files right? For snapshots as currently planned (i.e. stored in directories), users will always interact with them via the GUI or `rdctl`, right?

I'd put a `snapshots` directory right next to the `extensions` directory and the `lima` directory, and just not delete it during factory reset. This has the added benefit of making it easy to do a true factory reset if desired, by deleting the parent directory. I could see `$HOME/.local/share/rancher-desktop-snapshots` working as well.

It's your call I guess :shrug:

**Comment by jandubois** [comment]
I don't agree with the "don't use spaces in filenames" argument, but I agree that `$HOME/Rancher Desktop Snapshots` does stick out like a sore thumb due to its length.

I still feel uneasy about sticking it into the "app data" directory, because we currently delete it as part of a factory reset. But if we change it to only delete specific files and subdirectories, then I guess it will be fine.

So:

* macOS: `~/Library/Application Support/rancher-desktop/snapshots`
* Linux: `~/.local/share/rancher-desktop/snapshots`
* Windows: `%LOCALAPPDATA%\rancher-desktop\snapshots`

Does that work for you @adamkpickering? I can update the epic on top if we agree on this.

We will need to figure out an "Uninstall" mechanism to reclaim this space. But the UI will allow you to select all snapshots, and then delete the selected snapshots, so it will already be possible to clean it up from the UI.

**Comment by jandubois** [comment]
I'm going to update the snapshot directory section with a request to exclude it from Time Machine configuration by default, the same way we already do this for the Lima directory: https://github.com/rancher-sandbox/rancher-desktop/blob/f7a808823241102b8293a9c7166ff94726aacdf6/pkg/rancher-desktop/backend/lima.ts#L744-L750

**Comment by adamkpickering** [comment]
> Does that work for you @adamkpickering? I can update the epic on top if we agree on this.

Sounds good to me :+1:

The cleanest way I can think of is to have an option in factory reset to delete snapshots. But that conversation is for another time I think. The implementation of whatever solution we come up shouldn't be a lot of work either.

**Comment by adamkpickering** [comment]
@jandubois, @rak-phillip and I just had somewhat of a chat about snapshots. It's not very organized, but here is what we talked about:

Is there a copy system call (on any system) that we can use to make snapshot creation/restoration faster? If so, and if it is considerably faster than just reading the source and writing the destination, we want to use it, even if it means we cannot report copy progress to the user.

Otherwise, we want to communicate progress to the user when doing the plain copy. If we do communicate progress, it shall be via newline-delimited JSON objects written to stdout. Progress (at least, the kind of progress that is used by the UI) will be enabled by a flag, so that we don't spam the user with gibberish when they create snapshots on the command line.

Behavior when snapshot creation fails is simple: simply delete the snapshot directory. Behavior when snapshot restoration fails is more complex:
- report a message as to why it failed
- back up current settings.json and (if applicable) override.yaml
- factory reset & shut down
- restore settings.json and (if applicable) override.yaml in order to avoid first run dialog
- start Rancher Desktop again

We need to report useful error messages to the user. We are not sure how this will work. Some error conditions we anticipate:
- out of disk space
- broken symlink

Deleting large files on Unixes is fast. However, it may be slow on Windows. We need to test this. How long does it take to delete a 100GB file on Windows?

How can we prevent listing (and thus possibly re-deleting) a snapshot file after it has been deleted?
- The presence of metadata.json indicates a valid (i.e. not deleted, completely created) snapshot.
- metadata.json gets deleted first, and only then do we do the recursive delete of the snapshot directory.
- `rdctl snapshot list` should not list snapshots that do not have a metadata.json.

What happens if the system crashes or reboots in the middle of snapshot creation or deletion? We end up with an incomplete snapshot directory. This directory may occupy a lot of disk space, so we need to clean it up - how would we do this? One idea discussed here was the creation of a PID file inside the snapshots directory. The existence of this file indicates that a snapshot creation/deletion is in progress, and what the PID of the `rdctl` process performing this operation is. That way, if we see that there is a PID file, but no corresponding process, we can delete any incomplete snapshots. This may not work on Windows though, since Windows may re-use PIDs?

**Comment by adamkpickering** [comment]
I finally got a chance to do some more research about whether the platforms we are working on have copy system calls. The results of my research are below:

Linux: no copy system call that I can find (other than maybe `ioctl_ficlone`, which we are already using)
macOS: no copy system call that I can find (other than maybe `clonefile`, which we are already using)
Windows: `CopyFile` and `CopyFileEx`; further documentation [here](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-copyfile)

**Comment by adamkpickering** [comment]
As suggested by @jandubois, I am going to write out my ideas here in order to clarify them, and to serve as a starting point for Friday's discussion about how the UI and backend will work together regarding snapshots.

### How do the UI and backend work together regarding snapshots?

It should be possible to perform snapshot operations (create, restore, list, delete) from both `rdctl` and from the UI. In order to avoid duplication of functionality, we decided to put all of the functionality into `rdctl`. For example, if a user wants to create a snapshot via the command line, they would do so by invoking `rdctl snapshot create`. And if the user wants to create a snapshot using the UI, they click a button (I'm not sure of the details here) that causes the main process to call `rdctl snapshot create`. The only way the UI can perform snapshot-related operations is via an `rdctl snapshot ...` command. These commands should provide whatever information and functionality the UI needs to provide a good experience to the user.

At this point some questions you may have are:
- How does `rdctl` talk to the main process?
- How does `rdctl` stop/start the VM before/after creation/restoration of a snapshot?
- How does `rdctl` prevent the user from making changes to Rancher Desktop while a snapshot is being created?

Generally speaking, the answer to each one of these questions is "via the main process' HTTP API". I propose creating, or using if they are already present, the following endpoints on the main process:

- `GET /v1/backend_state`
- `GET /v1/settings`
- `PUT /v1/start_backend`
- `PUT /v1/stop_backend`
- `PUT /v1/snapshot_settings`
- `PUT /v1/snapshot_operation_in_progress`

Let's look at the step-by-step working of an `rdctl snapshot create` invokation:

1. Place main process in "snapshot operation in progress" state via `PUT /v1/snapshot_operation_in_progress`
2. Request backend state `STOPPED` via `PUT /v1/desired_backend_state`
3. Poll `GET /v1/backend_state` until the backend state is `STOPPED`
4. Create the snapshot
6. Request backend state `STARTED` via `PUT /v1/desired_backend_state`
7. Poll `GET /v1/backend_state` until the backend state is `STARTED`
8. Place main process in "snapshot operation not in progress" state via `PUT /v1/snapshot_operation_in_progress`

The purpose of the `/v1/snapshot_operation_in_progress` endpoint is to control the value of a variable (let's say `snapshotOperationInProgress`) in background.ts that tracks whether we are in the middle of a "snapshot operation". We would also have events for propagating changes to this variable for the purposes of the UI. If these events and this variable indicate that a snapshot operation is in progress, parts of RD that use them should disallow changes that could cause havoc with the snapshot operation. For example, it must be impossible to open the preferences dialog while a snapshot operation is in progress. Likewise, it must be impossible to do a `rdctl set ...` command.

Regarding the `/v1/start_backend` and `/v1/stop_backend` endpoints, this is fairly self-explanatory. They are used to request that the backend start or stop. This is necessary while creating or restoring a snapshot: if the VM is changed while it is being copied to the snapshot directory, it is possible that it will get into an unworkable state.

### About Settings

There is another wrinkle to be aware of: restoring settings. Originally, it was conceived that snapshots would be as simple as shutting down the VM and copying the files that represent the state of Rancher Desktop into a snapshots directory. Restoring those files would then be as simple as copying them back into their places. This works for most files, but settings.json is a special case.

This is because the main process does not monitor settings.json for changes after it loads the settings on startup. So if we are restoring from a snapshot while Rancher Desktop is running, I can think of two options that are available to us:

1. Get it to re-load settings in some way (in practice, this means a PUT API endpoint).
2. Send the settings we are restoring to to an API endpoint.

I prefer option 2 because the interface is smaller, and we already have something like it (`PUT /v1/settings`) in the API.

Of course, if Rancher Desktop is not running, we have no API to work with. We also don't have the problem of getting Rancher Desktop to reload the settings. In this case, copying the relevant files back and forth between their working locations and their snapshot locations works just fine. settings.json is no exception.

You may wonder why we have a `PUT /v1/snapshot_settings` when we already have `PUT /v1/settings`. This is because `PUT /v1/settings` has a different purpose, and accordingly it has some behaviors that are undesirable for restoring settings:
- It throws an error if the received settings object contains a value that differs from the corresponding value the parsed locked deployment profile. When restoring settings, we instead want to apply any differing values from the locked deployment profile to the received settings object. Unless I am mistaken, this should work the same as parsing settings on startup.
- It immediately restarts the backend if it detects that the received settings require this. We do not want the backend to be restarted until we have finished doing all of the things associated with restoring a snapshot, because doing so may result in an error. Only `rdctl` knows when this process is complete, so it is the only thing that should trigger a backend restart.

During snapshot creation I intend to use `GET /v1/settings` in order to get the settings object that goes into the snapshot. You may wonder why this is. I don't actually have a good argument for this, other than the following:
- It is symmetrical with snapshot restoration, which uses the API to restore settings as described above. Because it is symmetrical it is less surprising.
- It makes `rdctl` closer to being the client that I believe it should be.


### Some Things to Discuss

I want to end this by noting a couple of points that are relevant to the UI:

- Will the `snapshotOperationInProgress` variable and events that are emitted when it changes be enough to disable whichever parts of the UI should be disabled when we are doing a snapshot operation?
- We need to come up with a well-defined interface between `rdctl snapshot ...` commands and the UI. I can think of the following questions:
  - How should errors work? I imagine some kind of JSON object written to stderr would be best.
  - How do we communicate progress of long-running operations? In previous conversations we agreed that newline-delimited JSON objects would work well. Note that this is a lower priority discussion since it is not needed for 1.11.0.
- We need options on the `rdctl snapshot ...` commands that will facilitate their use by the UI. I can see these options taking several forms:
  - We shouldn't print JSON errors or JSON progress when the user invokes one of these commands under normal circumstances. So we need some option that causes these to be printed. `--for-main-process`?
  - We should add `--json` to `rdctl snapshot list`, which would cause it to output the snapshots as an array of JSON objects. This way we can avoid doing unreliable text parsing in the main process.

I think that sums up my thoughts. Let me know what you think!

**Comment by jandubois** [comment]
Just some quick initial feedback in case you have time to read before our meeting:

#### `PUT /v1/backend_state`

This can replace both `PUT /v1/start_backend` and `PUT /v1/stop_backend`. The desired state should be a parameter in the payload (e.g. `{"state": "RUNNING"}`), and not part of the API name. Seems like a logical companion to `GET /v1/backend_state`. I might even want to drop the `_state` suffix because every get/put is manipulating state.

I think this API should also include a `locked` parameter (e.g. `{"state": "STOPPED", "locked": "Snapshot creation in progress"}`, that specifies that the backend will be in a locked/unlocked state after it has been stopped/started. I believe this will get rid of the need for `PUT /v1/snapshot_operation_in_progress`. I prefer a more generic name, in case we want to use the same mechanism for something else in the future.

The locked state must be reflected in the filesystem: If the app is started while a snapshot is already in progress, the app needs to detect this automatically and stay in locked state.

This also means that `rdctl` needs to unlock the app via the API, even if the locking was performed just via a file system semaphore because the app was not yet running.

Finally, we need a mechanism to manually unlock the backend in case the `rdctl snapshot` command has died in the middle of taking a snapshot (killed, machine has crashed, etc), and the GUI needs to display a hint how to unlock when it starts up and the lock is already in place. Could be as simple as `rdctl unlock`.

## Settings

I think the regular app startup should invoke `PUT /v1/backend_state` with `{"state": "RUNNING"}` (unless the VM is currently locked). This code should then read the settings and deployment profile.

This means that the `rdctl snapshot` code doesn't have to do anything special to restore the settings beyond updating `settings.json`; they will be processed using the regular code path whenever the VM is started.

**Comment by adamkpickering** [comment]
I agree with everything in the comment, except:

> I think the regular app startup should invoke PUT /v1/backend_state with {"state": "RUNNING"} (unless the VM is currently locked). This code should then read the settings and deployment profile.

 If we do this, we are combining two operations into one:

- saying that we want the backend to be running
- loading/reloading the settings

I am worried this would be surprising. Nobody expects settings to be reloaded when they express that they want the backend to be running, right? I think it would be clearer if there was one endpoint for each operation.

That's assuming, of course, that we don't want to set settings via the API as I described above. It certainly isn't ideal to have the very similar endpoints of `PUT /v1/settings` and `PUT /v1/snapshot_settings`.

The more I think about it, the more I like the idea of having an endpoint that reloads settings when invoked. I think it is the shortest path to achieving the functionality we want.

**Comment by jandubois** [comment]
@adamkpickering Please create a follow-up epic for the 1.12 milestone and close this issue once all still open stories and information have been transferred to the new epic!
